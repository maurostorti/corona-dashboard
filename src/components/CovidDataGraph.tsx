import { LineChart, Line, XAxis, YAxis, Tooltip, Legend } from 'recharts'
import '../App.css'
import { DailyDataType } from '../utils/api-types'

const options = {
  confirmed: {
    label: 'Confirmed cases',
    stroke: '#8884d8',
  },
  recovered: {
    label: 'Recovered cases',
    stroke: '#82ca9d',
  },
  deaths: {
    label: 'Deaths',
    stroke: '#7D3B37',
  },
}

type CovidDataGraphProps = {
  graphData: DailyDataType[]
  graphSelected: string
  startDateIndex: number
  endDateIndex: number
}
export default function CovidDataGraph({
  graphData,
  graphSelected,
  startDateIndex,
  endDateIndex,
}: CovidDataGraphProps) {
  // Current selected options
  const opts: (keyof typeof options)[] =
    graphSelected === 'all' ? (Object.keys(options) as any[]) : [graphSelected]

  // Lines to be rendered
  const lines = opts.map((opt) => {
    return (
      <Line
        key={opt}
        type='monotone'
        dataKey={`${opt}.total`}
        name={options[opt].label as any}
        stroke={options[opt].stroke as any}
        dot={false}
      />
    )
  })

  return (
    <LineChart
      width={730}
      height={400}
      data={graphData.slice(startDateIndex, endDateIndex)}
      margin={{ top: 5, right: 30, left: 20, bottom: 5 }}
    >
      <XAxis dataKey='reportDate' tick={false} />
      <YAxis />
      <Tooltip />
      <Legend />
      {lines}
    </LineChart>
  )
}
