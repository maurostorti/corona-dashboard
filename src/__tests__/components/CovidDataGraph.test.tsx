import { render, screen } from '@testing-library/react'
import CovidDataGraph from '../../components/CovidDataGraph'

const graphDataTest = [
  {
    confirmed: {
      total: 10,
    },
    deaths: {
      total: 5,
    },
    recovered: {
      total: 2,
    },
    reportDate: '2020-10-30',
  },
  {
    confirmed: {
      total: 1,
    },
    deaths: {
      total: 50,
    },
    recovered: {
      total: 23,
    },
    reportDate: '2021-01-20',
  },
]

test('it renders correctly confirmed', () => {
  const { asFragment } = render(
    <CovidDataGraph
      graphData={graphDataTest}
      graphSelected='all'
      startDateIndex={0}
      endDateIndex={graphDataTest.length - 1}
    />,
  )
  expect(asFragment()).toMatchSnapshot()
})

test('it renders legend all', () => {
  render(
    <CovidDataGraph
      graphData={graphDataTest}
      graphSelected='all'
      startDateIndex={0}
      endDateIndex={graphDataTest.length - 1}
    />,
  )
  const confirmedLegend = screen.getByText('Confirmed cases')
  expect(confirmedLegend).toBeInTheDocument()

  const recoveredLegend = screen.getByText('Recovered cases')
  expect(recoveredLegend).toBeInTheDocument()

  const deathsLegend = screen.getByText('Deaths')
  expect(deathsLegend).toBeInTheDocument()
})

test('it renders legend confirmed', () => {
  render(
    <CovidDataGraph
      graphData={graphDataTest}
      graphSelected='confirmed'
      startDateIndex={0}
      endDateIndex={graphDataTest.length - 1}
    />,
  )
  const confirmedLegend = screen.queryByText('Confirmed cases')
  expect(confirmedLegend).toBeInTheDocument()

  const recoveredLegend = screen.queryByText('Recovered cases')
  expect(recoveredLegend).not.toBeInTheDocument()

  const deathsLegend = screen.queryByText('Deaths')
  expect(deathsLegend).not.toBeInTheDocument()
})

test('it renders legend recovered', () => {
  render(
    <CovidDataGraph
      graphData={graphDataTest}
      graphSelected='recovered'
      startDateIndex={0}
      endDateIndex={graphDataTest.length - 1}
    />,
  )
  const confirmedLegend = screen.queryByText('Confirmed cases')
  expect(confirmedLegend).not.toBeInTheDocument()

  const recoveredLegend = screen.queryByText('Recovered cases')
  expect(recoveredLegend).toBeInTheDocument()

  const deathsLegend = screen.queryByText('Deaths')
  expect(deathsLegend).not.toBeInTheDocument()
})

test('it renders legend deaths', () => {
  render(
    <CovidDataGraph
      graphData={graphDataTest}
      graphSelected='deaths'
      startDateIndex={0}
      endDateIndex={graphDataTest.length - 1}
    />,
  )
  const confirmedLegend = screen.queryByText('Confirmed cases')
  expect(confirmedLegend).not.toBeInTheDocument()

  const recoveredLegend = screen.queryByText('Recovered cases')
  expect(recoveredLegend).not.toBeInTheDocument()

  const deathsLegend = screen.queryByText('Deaths')
  expect(deathsLegend).toBeInTheDocument()
})
