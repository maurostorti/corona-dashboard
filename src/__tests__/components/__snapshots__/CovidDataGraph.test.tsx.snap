// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly confirmed 1`] = `
<DocumentFragment>
  <div
    class="recharts-wrapper"
    style="position: relative; cursor: default; width: 730px; height: 400px;"
  >
    <svg
      class="recharts-surface"
      height="400"
      version="1.1"
      viewBox="0 0 730 400"
      width="730"
    >
      <defs>
        <clippath
          id="recharts1-clip"
        >
          <rect
            height="360"
            width="620"
            x="80"
            y="5"
          />
        </clippath>
      </defs>
      <g
        class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
      >
        <line
          class="recharts-cartesian-axis-line"
          fill="none"
          height="30"
          orientation="bottom"
          stroke="#666"
          type="category"
          width="620"
          x="80"
          x1="80"
          x2="700"
          y="365"
          y1="365"
          y2="365"
        />
        <g
          class="recharts-cartesian-axis-ticks"
        />
      </g>
      <g
        class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
      >
        <line
          class="recharts-cartesian-axis-line"
          fill="none"
          height="360"
          orientation="left"
          stroke="#666"
          type="number"
          width="60"
          x="20"
          x1="80"
          x2="80"
          y="5"
          y1="5"
          y2="365"
        />
        <g
          class="recharts-cartesian-axis-ticks"
        >
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="360"
              orientation="left"
              stroke="#666"
              type="number"
              width="60"
              x="20"
              x1="74"
              x2="80"
              y="5"
              y1="365"
              y2="365"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="360"
              orientation="left"
              stroke="none"
              text-anchor="end"
              type="number"
              width="60"
              x="72"
              y="365"
            >
              <tspan
                dy="0.355em"
                x="72"
              >
                0
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="360"
              orientation="left"
              stroke="#666"
              type="number"
              width="60"
              x="20"
              x1="74"
              x2="80"
              y="5"
              y1="275"
              y2="275"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="360"
              orientation="left"
              stroke="none"
              text-anchor="end"
              type="number"
              width="60"
              x="72"
              y="275"
            >
              <tspan
                dy="0.355em"
                x="72"
              >
                3
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="360"
              orientation="left"
              stroke="#666"
              type="number"
              width="60"
              x="20"
              x1="74"
              x2="80"
              y="5"
              y1="185"
              y2="185"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="360"
              orientation="left"
              stroke="none"
              text-anchor="end"
              type="number"
              width="60"
              x="72"
              y="185"
            >
              <tspan
                dy="0.355em"
                x="72"
              >
                6
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="360"
              orientation="left"
              stroke="#666"
              type="number"
              width="60"
              x="20"
              x1="74"
              x2="80"
              y="5"
              y1="95"
              y2="95"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="360"
              orientation="left"
              stroke="none"
              text-anchor="end"
              type="number"
              width="60"
              x="72"
              y="95"
            >
              <tspan
                dy="0.355em"
                x="72"
              >
                9
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="360"
              orientation="left"
              stroke="#666"
              type="number"
              width="60"
              x="20"
              x1="74"
              x2="80"
              y="5"
              y1="5"
              y2="5"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="360"
              orientation="left"
              stroke="none"
              text-anchor="end"
              type="number"
              width="60"
              x="72"
              y="5"
            >
              <tspan
                dy="0.355em"
                x="72"
              >
                12
              </tspan>
            </text>
          </g>
        </g>
      </g>
      <g
        class="recharts-layer recharts-line"
      >
        <g
          class="recharts-layer recharts-line-dots"
        >
          <circle
            class="recharts-dot recharts-line-dot"
            cx="390"
            cy="64.99999999999999"
            fill="#fff"
            height="360"
            name="Confirmed cases"
            points="[object Object]"
            r="3"
            stroke="#8884d8"
            stroke-width="1"
            type="monotone"
            width="620"
          />
        </g>
      </g>
      <g
        class="recharts-layer recharts-line"
      >
        <g
          class="recharts-layer recharts-line-dots"
        >
          <circle
            class="recharts-dot recharts-line-dot"
            cx="390"
            cy="305"
            fill="#fff"
            height="360"
            name="Recovered cases"
            points="[object Object]"
            r="3"
            stroke="#82ca9d"
            stroke-width="1"
            type="monotone"
            width="620"
          />
        </g>
      </g>
      <g
        class="recharts-layer recharts-line"
      >
        <g
          class="recharts-layer recharts-line-dots"
        >
          <circle
            class="recharts-dot recharts-line-dot"
            cx="390"
            cy="214.99999999999997"
            fill="#fff"
            height="360"
            name="Deaths"
            points="[object Object]"
            r="3"
            stroke="#7D3B37"
            stroke-width="1"
            type="monotone"
            width="620"
          />
        </g>
      </g>
    </svg>
    <div
      class="recharts-legend-wrapper"
      style="position: absolute; width: 680px; height: auto; left: 20px; bottom: 5px;"
    >
      <ul
        class="recharts-default-legend"
        style="padding: 0px; margin: 0px; text-align: center;"
      >
        <li
          class="recharts-legend-item legend-item-0"
          style="display: inline-block; margin-right: 10px;"
        >
          <svg
            class="recharts-surface"
            height="14"
            style="display: inline-block; vertical-align: middle; margin-right: 4px;"
            version="1.1"
            viewBox="0 0 32 32"
            width="14"
          >
            <path
              class="recharts-legend-icon"
              d="M0,16h10.666666666666666
            A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
            H32M21.333333333333332,16
            A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16"
              fill="none"
              stroke="#8884d8"
              stroke-width="4"
            />
          </svg>
          <span
            class="recharts-legend-item-text"
            style="color: rgb(136, 132, 216);"
          >
            Confirmed cases
          </span>
        </li>
        <li
          class="recharts-legend-item legend-item-1"
          style="display: inline-block; margin-right: 10px;"
        >
          <svg
            class="recharts-surface"
            height="14"
            style="display: inline-block; vertical-align: middle; margin-right: 4px;"
            version="1.1"
            viewBox="0 0 32 32"
            width="14"
          >
            <path
              class="recharts-legend-icon"
              d="M0,16h10.666666666666666
            A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
            H32M21.333333333333332,16
            A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16"
              fill="none"
              stroke="#82ca9d"
              stroke-width="4"
            />
          </svg>
          <span
            class="recharts-legend-item-text"
            style="color: rgb(130, 202, 157);"
          >
            Recovered cases
          </span>
        </li>
        <li
          class="recharts-legend-item legend-item-2"
          style="display: inline-block; margin-right: 10px;"
        >
          <svg
            class="recharts-surface"
            height="14"
            style="display: inline-block; vertical-align: middle; margin-right: 4px;"
            version="1.1"
            viewBox="0 0 32 32"
            width="14"
          >
            <path
              class="recharts-legend-icon"
              d="M0,16h10.666666666666666
            A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
            H32M21.333333333333332,16
            A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16"
              fill="none"
              stroke="#7D3B37"
              stroke-width="4"
            />
          </svg>
          <span
            class="recharts-legend-item-text"
            style="color: rgb(125, 59, 55);"
          >
            Deaths
          </span>
        </li>
      </ul>
    </div>
    <div
      class="recharts-tooltip-wrapper"
      style="transform: translate(undefinedpx, undefinedpx); pointer-events: none; visibility: hidden; position: absolute; top: 0px; left: 0px;"
    >
      <div
        class="recharts-default-tooltip"
        style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
      >
        <p
          class="recharts-tooltip-label"
          style="margin: 0px;"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
